sherpa/bin/offline_server.py

```diff
@@ -0,0 +1,388 @@
+            CUDA_VISBILE_DEVICES to achieve this.
+          batch_size:
+            Max batch size for inference.
+          max_wait_ms:
+            Max wait time in milliseconds in order to build a batch of
+            `batch_size`.
+          feature_extractor_pool_size:
+            Number of threads to create for the feature extractor thread pool.
+          nn_pool_size:
+            Number of threads for the thread pool that is used for NN
+            computation and decoding.
+        """
+        self.feature_extractor = self._build_feature_extractor()
+        self.nn_models = self._build_nn_model(nn_model_filename, num_device)
+
+        assert nn_pool_size > 0
+
+        self.feature_extractor_pool = ThreadPoolExecutor(
+            max_workers=feature_extractor_pool_size,
+            thread_name_prefix="feature",
+        )
+        self.nn_pool = ThreadPoolExecutor(
+            max_workers=nn_pool_size,
+            thread_name_prefix="nn",
+        )
+
+        self.feature_queue = asyncio.Queue()
+
+        self.sp = spm.SentencePieceProcessor()
+        self.sp.load(bpe_model_filename)
+
+        self.counter = 0
+
+        self.max_wait_ms = max_wait_ms
+        self.batch_size = batch_size
+
+    def _build_feature_extractor(self):
+        """Build a fbank feature extractor for extracting features.
+
+        TODO:
+          Pass the options as arguments
+        """
+        opts = kaldifeat.FbankOptions()
+        opts.device = "cpu"  # Note: It also supports CUDA, e.g., "cuda:0"
+        opts.frame_opts.dither = 0
+        opts.frame_opts.snip_edges = False
+        opts.frame_opts.samp_freq = 16000
+        opts.mel_opts.num_bins = 80
+
+        fbank = kaldifeat.Fbank(opts)
+
+        return fbank
+
+    def _build_nn_model(
+        self, nn_model_filename: str, num_device: int
+    ) -> List[RnntModel]:
+        """Build a torch script model for each given device.
+
+        Args:
+          nn_model_filename:
+            The path to the torch script model.
+          num_device:
+            Number of devices to use for NN computation and decoding.
+            If it is 0, then only use CPU and it returns a model on CPU.
+            If it is positive, it create a model for each device and returns
+            them.
+        Returns:
+          Return a list of torch script models.
+        """
+        if num_device < 1:
+            model = RnntModel(
+                filename=nn_model_filename,
+                device="cpu",
+                optimize_for_inference=False,
+            )
+            return [model]
+
+        ans = []
+        for i in range(num_device):
+            device = torch.device("cuda", i)
+            model = RnntModel(
+                filename=nn_model_filename,
+                device=device,
+                optimize_for_inference=False,
+            )
+            ans.append(model)
+
+        return ans
+
+    async def loop(self, port: int):
```

Comment:
There is the entry point to the server. It invokes `self.handle_connection` on each new connection.